### DISCUSSION TITLE
Custom OTP endpoint verification / Verify `TokenHash`

### DISCUSSION THREAD

**Discussion by rawnly** [supabase/supabase#12627] [Questions]
Hello, I'm using `magiLink` to handle log in to my application, but I want the user to be able to log in via OTP too. 

So I built a custom route in my nextjs application to validate OTP from query params. How do I decode the `{{ .TokenHash }}` once received (maybe I misunderstood how to use it properly) 

I'm currently on nextjs with authentication managed inside the middleware. Using the default magicLink cause the app to reject the user at the first attempt (i think it needs the page to load to set the cookie right?) 

So is creating some kind of `/otp` unauthenticated route, just to set the cookie and redirect, the only alternative?

The docs says:

> Contains a hashed version of the `{{ .Token }}`. This is useful for constructing your own email link in the email template

I think I'm missing something and I can't find anything on Github / Documentation / Discussions :|

**Comment by PrabhdeepSingh** [comment]
I'm experiencing a similar situation. 
`supabase.auth.verifyOtp` doesn't accept the `{{ .TokenHash }}`, and I need to figure out a way to decode it.

Did you have any luck finding a solution?

**Reply by rawnly** [reply]
I opened a ticket and got the following response

 ![image](https://user-images.githubusercontent.com/16429579/235379997-c9383441-882f-4fce-9968-02a0faa5ca57.jpeg)

https://supabase.com/docs/guides/auth/auth-email-templates#limitations

**Comment by atothewest** [comment]
If anyone stumbles upon this and is interested in an answer, you can use the `.TokenHash` instead of the `.Token` in the `verifyOtp` method:

```
await supabase.auth.verifyOtp({ token_hash: tokenHash, type: 'email' });
```

See:
https://supabase.com/docs/reference/javascript/auth-verifyotp
